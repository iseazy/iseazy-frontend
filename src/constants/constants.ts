export const constants = {
    gameTitle: 'MeMemory',
    buttons: {
        accept: 'Aceptar',
        cancel: 'Cancelar',
        ranking: 'Ranking',
        start: 'Comenzar'
    },
    ranking: {
        title: 'Ranking',
        noData: 'No hay datos para mostrar',
        goBack: 'Volver al Menú',
        header: {
            position: 'Pos',
            name: 'Nombre',
            time: 'Tiempo',
            attempts: 'Mov'
        }
    },
    modal: {
        title: '¡Completado!',
        attempts: 'Movimientos',
        insertYourName: 'Ingresa tu nombre',
    }
}