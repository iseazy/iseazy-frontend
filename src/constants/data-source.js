export const CARDS_SRC = [
    'images/card-001.png',
    'images/card-002.png',
    'images/card-003.png',
    'images/card-004.png',
    'images/card-005.png',
    'images/card-006.png',
    'images/card-007.png',
    'images/card-008.png',
    'images/card-009.png'
];
