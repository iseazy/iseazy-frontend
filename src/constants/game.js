export const GAME_STATE = {
    IDLE: 'idle',
    PLAY: 'play',
    END: 'end'
}

export const GAME_ACTIONS = {
    START: 'game_start',
    END: 'game_end',
    CLICK: 'game_click',
    UPDATE: 'game_update'
}

export const INITIAL_STATE = {
    gameState: GAME_STATE.IDLE,
    cards: [],
    startTime: 0,
    totalTime: 0
}

export const REPEATED_CARDS =  2;

export const GAME_UPDATE_DELAY = 500;
