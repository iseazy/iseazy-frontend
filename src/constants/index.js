export { CARDS_SRC } from './data-source';

export { GAME_ACTIONS, GAME_STATE, GAME_UPDATE_DELAY, INITIAL_STATE, REPEATED_CARDS } from './game';
