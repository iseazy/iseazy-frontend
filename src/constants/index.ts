import { Item } from '../entities/item.entity'

export const COVER_IMAGE = '/images/cover.svg'

export const AVAILABLE_ITEMS: Item[] = [
  {
    key: 'cat',
    image: '/images/cat.png',
  },
  {
    key: 'crying',
    image: '/images/crying.png',
  },
  {
    key: 'doge',
    image: '/images/doge.png',
  },
  {
    key: 'girl',
    image: '/images/girl.png',
  },
  {
    key: 'patricio',
    image: '/images/patricio.png',
  },
  {
    key: 'poker-face',
    image: '/images/poker-face.png',
  },
  {
    key: 'smile',
    image: '/images/smile.png',
  },
  {
    key: 'suspect',
    image: '/images/suspect.png',
  },
  {
    key: 'think',
    image: '/images/think.png',
  },
]
