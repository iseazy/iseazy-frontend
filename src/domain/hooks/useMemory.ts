import { useState, useEffect } from 'react'
import CardsFactory from '../memory/cards'
import {
  calculateStatus,
  selectCard as select,
  deselectCards,
  isClickableCard,
} from '../memory/logic'

interface State {
  time: {
    start: number | null
    end: number | null
  }
  selectedCardIds: String[]
  cards: ReturnType<typeof CardsFactory>
}

const initialState = (): State => {
  return {
    time: {
      start: null,
      end: null,
    },
    selectedCardIds: [],
    cards: CardsFactory(),
  }
}

function useMemory() {
  const [state, setState] = useState(initialState())
  // derived state
  const status = calculateStatus({
    selectedCardIds: state.selectedCardIds,
    cards: state.cards,
  })

  useEffect(() => {
    if (status !== 'unmatch') return
    const timer = setTimeout(() => {
      setState((currentState) => deselectCards(currentState))
    }, 900)
    return () => clearTimeout(timer)
  }, [status])

  const selectCard = (id: string) => {
    if (isClickableCard({ id, status, cards: state.cards }))
      setState((currentState) => select(id, currentState))
  }

  const restartGame = () => {
    setState(initialState())
  }

  return {
    time: state.time,
    status,
    cards: state.cards,
    selectCard,
    restartGame,
  }
}

export type { State }
export default useMemory
