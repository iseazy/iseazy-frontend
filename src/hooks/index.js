export { useGame } from './use-game';
