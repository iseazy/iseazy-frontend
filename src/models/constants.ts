export const TYPE_BUTTONS  = {
    PRIMARY: 'btn--primary',
    SECONDARY: 'btn--secondary'
};

export const TOTAL_CARDS = 18;
export const FINISHGAME = "FINISHGAME";
export const STARTGAME = "STARTGAME";
export const SOLVEDMATCH = "SOLVEDMATCH";
export const FLIPPEDCARD = "FLIPPEDCARD";