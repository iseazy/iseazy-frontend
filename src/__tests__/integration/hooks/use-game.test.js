import { renderHook, act } from '@testing-library/react-hooks';
import { useGame } from '../../../hooks';
import { GAME_STATE } from '../../../constants';

jest.mock('../../../constants', () => ({
    ...jest.requireActual('../../../constants'),
    CARDS_SRC: [ 'sourceA', 'sourceB' ]
}));

describe('useGame()', () => {
    it('should render hook with an initial state', () => {
        const { result } = renderHook(() => useGame());

        expect(result.current).toMatchSnapshot();
    });

    it('should respond to start game event', () => {
        const { result } = renderHook(() => useGame());

        act(() => { 
            result.current.start();
        });

        expect(result.current.cards.length).toBeGreaterThan(0);
    });

    it('should be able to complete the game', async () => {
        const { result , waitForValueToChange, waitForNextUpdate} = renderHook(() => useGame());

        act(() => { result.current.start(); });

        const [a, b] = result.current.cards.filter(({ src }) => src === 'sourceA');
        const [c, d] = result.current.cards.filter(({ src }) => src === 'sourceB');

        act(() => { result.current.clickCard(a.id); });
        act(() => { result.current.clickCard(b.id); });

        await waitForValueToChange(() => { return result.current.cards; });

        act(() => { result.current.clickCard(c.id); });
        act(() => { result.current.clickCard(d.id); });

        await waitForValueToChange(() => { return result.current.cards; });

        expect(result.current.gameState).toBe(GAME_STATE.END);
    });

    it('should be able to fail pairing', async () => {
        const { result , waitForValueToChange, waitForNextUpdate} = renderHook(() => useGame());

        act(() => { result.current.start(); });

        const [a, b] = result.current.cards.filter(({ src }) => src === 'sourceA');
        const [c, d] = result.current.cards.filter(({ src }) => src === 'sourceB');

        act(() => { result.current.clickCard(a.id); });
        act(() => { result.current.clickCard(c.id); });

        await waitForValueToChange(() => { return result.current.cards; });

        expect(result.current.gameState).toBe(GAME_STATE.PLAY);
    });
});
