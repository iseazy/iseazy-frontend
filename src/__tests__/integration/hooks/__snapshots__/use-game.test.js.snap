// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGame() should render hook with an initial state 1`] = `
Object {
  "cards": Array [],
  "clickCard": [Function],
  "gameState": "idle",
  "start": [Function],
  "totalTime": 0,
}
`;
